man/mtbl_merge.1
'\" t
.\"     Title: mtbl_merge
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 10/21/2021
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "MTBL_MERGE" "1" "10/21/2021" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mtbl_merge \- merge MTBL data from multiple input files into a single output file
.SH "SYNOPSIS"
.sp
User\-provided functions:
.sp
.nf
\fBtypedef void *
(*mtbl_merge_init_func)(void);\fR
.fi
.sp
.nf
\fBtypedef void
(*mtbl_merge_free_func)(void *clos);\fR
.fi
.sp
.nf
\fBtypedef void
(*mtbl_merge_func)(void *\fR\fB\fIclos\fR\fR\fB,
        const uint8_t *\fR\fB\fIkey\fR\fR\fB, size_t \fR\fB\fIlen_key\fR\fR\fB,
        const uint8_t *\fR\fB\fIval0\fR\fR\fB, size_t \fR\fB\fIlen_val0\fR\fR\fB,
        const uint8_t *\fR\fB\fIval1\fR\fR\fB, size_t \fR\fB\fIlen_val1\fR\fR\fB,
        uint8_t **\fR\fB\fImerged_val\fR\fR\fB, size_t *\fR\fB\fIlen_merged_val\fR\fR\fB);\fR
.fi
.sp
Command line tool:
.sp
.nf
\fBexport MTBL_MERGE_DSO="\fR\fB\fIlibexample\&.so\&.0\fR\fR\fB"\fR
\fBexport MTBL_MERGE_FUNC_PREFIX="\fR\fB\fIexample_merge\fR\fR\fB"\fR
\fBmtbl_merge\fR [\fB\-b\fR \fISIZE\fR] [\fB\-c\fR \fICOMPRESSION\fR] [\fB\-l\fR \fILEVEL\fR] \fIINPUT\fR [\fIINPUT\fR]\&... \fIOUTPUT\fR
.fi
.SH "DESCRIPTION"
.sp
\fBmtbl_merge\fR(1) is a command\-line driver for the \fBmtbl_merger\fR(3) interface\&. The \fBmtbl_merger\fR(3) interface requires a user\-provided merge function, which is loaded from a shared object whose filename is specified in the environment variable \fIMTBL_MERGE_DSO\fR\&.
.sp
The user\-provided merge function must have the same type as the \fImtbl_merge_func\fR function type given above in the synopsis\&. The symbol name of the merge function to be loaded from the user\-provided DSO will be constructed by appending "_func" to the string provided in the \fIMTBL_MERGE_FUNC_PREFIX\fR environment variable, which must be non\-empty\&.
.sp
Additionally, two optional functions may be provided: an "init" function whose symbol name is "_init" appended to the function prefix, and a "free" function whose symbol name is "_free" appended to the function prefix\&. If the "init" function exists, it will be called at the beginning, before any calls to the merge function, and the return value from the init function will be passed as the first argument to the merge function\&. If the "free" function exists, it will be called at the end, after any calls to the merge function, and its argument will be the return value of the "init" function\&.
.sp
The environment variable \fIMTBL_MERGE_BLOCK_SIZE\fR may optionally be set in order to configure the MTBL block size (in bytes) of the output file\&.
.SH "OPTIONS"
.PP
\fB\-b\fR \fISIZE\fR
.RS 4
The uncompressed data block size hint for the output file, in bytes\&. The default value if unspecified is 8192 bytes (8 kilobytes)\&.
.RE
.PP
\fB\-c\fR \fICOMPRESSION\fR
.RS 4
The compression algorithm to use for data blocks in the output file\&. The default value if unspecified is
\fBzlib\fR\&. See the
\fBmtbl_info\fR(1) manpage for the list of possible compression algorithms\&.
.RE
.PP
\fB\-l\fR \fILEVEL\fR
.RS 4
The numeric compression level passed to the compression algorithm\&. The default value and valid range depend on the chosen compression algorithm\&.
.RE
.SH "SEE ALSO"
.sp
\fBmtbl_info\fR(1), \fBmtbl_merger\fR(3)


ChangeLog
mtbl (1.3.0)

 * Fix mtbl_reload_now() on filesets with open iterators.
 * Add support for absolute pathnames in filesets
 * Make merge function an optional parameter for mergers and filesets to
   enable unmerged results.
 * Add dupsort function to sort unmerged results based on data.
 * Add filename filter option to filesets. This provides functionality similar
   to mtbl_fileset_partition() but resilient against fileset reloads.
 * Add mtbl_fileset_dup() to open an existing fileset with different options.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Fri, 29 Mar 2019 12:13:17 -0500

mtbl (1.2.1)

 * Fix libtool version number.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Fri May 25 17:49:08 2018 -0500

mtbl (1.2.0)

 * Prevent fileset reloading when the fileset has iterators open.
 * Add and document facility to disable reloading of filesets.
 * Defer initial load of fileset until the first operation on a fileset
   source.
 * Handle 32bit size_t overflows, failing with an assertion upon opening
   an mtbl_reader with an oversized data block.
 * Fix for systems with 32-bit size_t.
 * Add several unit tests for various libmtbl functions.

 -- Farsight Security, Inc. <software@farsightsecurity.com>  Fri, 25 May 2018 11:55:07 -0400

mtbl (1.1.1)

 * Fix iterator leak in mtbl_merger code.

 -- Chris Mikkelson <cmikk@fsi.io>  Wed, 2 Aug 2017 12:44:53 -0500

mtbl (1.1.0)

  * Fix default zlib compression level.
  * Add callback data (clos) parameter to mtbl_fileset_partition
    function and its callback. Early users of mtbl_fileset_partition
    will need to rewrite accordingly.
  * Use 64-bit offsets in blocks with more than 4G of data.
  * Fix undefined behavior when seeking past end of mtbl file.

mtbl (1.0.0)

 * Backwards-incompatible file format change to enable block sizes >4G.
 * Add support for zstd compression. This adds a new library dependency
   on libzstd.
 * Add mtbl_iter_seek function.
 * Add mtbl_fileset_partition function.
 * Breaks ABI for version 0.x.x.

mtbl (0.8.1)

 [ Robert Edmonds ]
 * Add portability for clock time.

 * Simplify and improve portability related to byte order primitives.

 [ Henry Stern ]
 * Fix assertion failure. If there is a broken mtbl file in the fileset
   then a NULL reader will be returned. This change checks for that
   error condition in fs_reinit_merger() and does not pass the empty
   reader onwards to mtbl_merger_add_source().

 -- Jeremy C. Reed <reed@fsi.io>  Wed, 23 Nov 2016 10:27:11 -0600

mtbl (0.8.0)

  [ Alexey Spiridonov ]
  * mtbl_reader(3): New reader getters, which expose the values stored
    in the "metadata" (formerly "trailer") at the end of MTBL files. For
    example: the number of bytes of source data in the keys & values is
    available via mtbl_metadata_bytes_keys() & mtbl_metadata_bytes_values().

  * mtbl_writer(3): Allow foreign data to be written to the beginning of a
    file before its file descriptor is passed to mtbl_writer_init_fd().

  [ Robert Edmonds ]
  * Add mtbl_verify(1) utility which verifies the embedded data and index
    block checksums in an MTBL file.

  * Stop keeping dup()'d copies of the file descriptors passed to
    mtbl_reader_init_fd(). POSIX does not require a process to keep an open
    file descriptor corresponding to an mmap()'d file. This change allows a
    process to open more MTBL files than the process file descriptor limit.

  * mtbl_dump(1): Add silent ("-s") option which omits the actual dump output.
    This is useful when benchmarking decompression performance.

  * Add LZ4/LZ4HC compression support. This adds a new library dependency on
    liblz4.

  * mtbl_merge(1): Add block size ("-b") and compression algorithm ("-c")
    options.

  * mtbl_fileset(3): Add mtbl_fileset_reload_now() function that
    instantaneously checks and, if necessary, reloads the fileset.

 -- Robert Edmonds <edmonds@fsi.io>  Fri, 28 Aug 2015 17:36:39 -0400

mtbl (0.7.0)

  * mtbl_reader(3): New reader option 'madvise_random' which may drastically
    improve performance on seek-heavy workloads. It can be enabled by the
    library caller with the mtbl_reader_options_set_madvise_random() function.
    It can also be globally force-enabled or force-disabled at runtime by
    setting the environment variable MTBL_READER_MADVISE_RANDOM to "1" or "0".

 -- Robert Edmonds <edmonds@fsi.io>  Wed, 19 Nov 2014 12:43:12 -0500

mtbl (0.6.0)

  * Fix assertion failures with highly compressed data caused by the use of a
    fixed size decompression buffer (Issue #1).

  * Fix small memory leak during initialization in the mtbl_merge utility.

  * Fix leak-on-error-bugs in the reader, sorter, and fileset interfaces,
    detected by static analysis.

  * Drop "-Wl,--as-needed" from LDFLAGS.

  * Begin versioning the library's symbols. (Based on ld-version-script.m4
    from gnulib.)

 -- Robert Edmonds <edmonds@fsi.io>  Wed, 21 May 2014 16:24:17 -0400

mtbl (0.5)

  * The COPYRIGHT, LICENSE, and README.md files are now distributed in the
    tarball. These files were inadvertently not included in the previous
    release.

  * The build system now properly detects big endian architectures. The
    previous release inadvertently omitted this check, causing a test suite
    failure.

  * Verify that the length of the MTBL input is long enough to read the MTBL
    header block. This prevents invalid reads in mtbl_reader_init_fd(). See
    Debian bug #716628 for details. (http://bugs.debian.org/716628).

  * Make sure to install the manpages if they are available even if the
    manpages are not being rebuilt.

 -- Robert Edmonds <edmonds@fsi.io>  Fri, 31 Jan 2014 18:01:13 -0500

mtbl (0.4)

  * Update copyright and license statements as a result of the transition from
    Internet Systems Consortium to Farsight Security.

  * Replace the "librsf" submodule with the "libmy" subtree. This includes a
    faster CRC32C implementation on supported x86-64 CPUs.


src/mtbl_merge.c
/*
 * Copyright (c) 2012, 2014-2016, 2019, 2021 by Farsight Security, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sys/time.h>
#include <assert.h>
#include <dlfcn.h>
#include <errno.h>
#include <inttypes.h>
#include <limits.h>
#include <locale.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <mtbl.h>

#include "libmy/getenv_int.h"
#include "libmy/ubuf.h"

#define DEFAULT_BLOCK_SIZE	8192
#define DEFAULT_COMPRESS_LEVEL	-1000
#define STATS_INTERVAL		1000000

static const char		*program_name;

static const char		*mtbl_output_fname;

static mtbl_compression_type	opt_compression_type	= MTBL_COMPRESSION_ZLIB;
static int			opt_compression_level	= DEFAULT_COMPRESS_LEVEL;
static size_t			opt_block_size		= DEFAULT_BLOCK_SIZE;

static const char		*merge_dso_path;
static const char		*merge_dso_prefix;

static mtbl_merge_init_func	user_func_init;
static mtbl_merge_free_func	user_func_free;
static mtbl_merge_func		user_func_merge;
static void			*user_clos;

static struct mtbl_merger	*merger;
static struct mtbl_writer	*writer;

static struct timespec		start_time;
static uint64_t			count;
static uint64_t			count_merged;

static void
usage(void)
{
	fprintf(stderr,
		"Usage: %s [-b <SIZE>] [-c <COMPRESSION>] [-l <LEVEL>] <INPUT> [<INPUT>...] <OUTPUT>\n"
		"\n"
		"Merges one or more MTBL input files into a single output file.\n"
		"Requires a merge function provided by the user at runtime via a DSO.\n"
		"See mtbl_merge(1) for details.\n"
		"\n"
		"<SIZE> is the uncompressed data block size hint, in bytes.\n"
		"The default value if unspecified is 8192 bytes (8 kilobytes).\n"
		"\n"
		"<COMPRESSION> is one of none, snappy, zlib, lz4, lz4hc, or zstd.\n"
		"The default compression type if unspecified is zlib.\n"
		"\n"
		"<LEVEL> is the numeric compression level passed to the compression algorithm.\n"
		"The default and valid range depend on the <COMPRESSION> parameter.\n"
		,
		program_name
	);
	exit(EXIT_FAILURE);
}

static void
my_timespec_get(struct timespec *now) {
	struct timeval tv;
	(void) gettimeofday(&tv, NULL);
	now->tv_sec = tv.tv_sec;
	now->tv_nsec = tv.tv_usec * 1000;
}

static void
my_timespec_sub(const struct timespec *a, struct timespec *b) {
	b->tv_sec -= a->tv_sec;
	b->tv_nsec -= a->tv_nsec;
	if (b->tv_nsec < 0) {
		b->tv_sec -= 1;
		b->tv_nsec += 1000000000;
	}
}

static double
my_timespec_to_double(const struct timespec *ts) {
	return (ts->tv_sec + ts->tv_nsec / 1E9);
}

static void
print_stats(void)
{
	struct timespec dur;
	double t_dur;

	my_timespec_get(&dur);
	my_timespec_sub(&start_time, &dur);
	t_dur = my_timespec_to_double(&dur);

	fprintf(stderr,
		"%s: wrote %'" PRIu64 " entries (%'" PRIu64 " merged) "
		"in %'.2f sec, %'d ent/sec, %'d merge/sec\n",
		program_name,
		count,
		count_merged,
		t_dur,
		(int) (count / t_dur),
		(int) (count_merged / t_dur)
	);
}

static void
merge_func(void *clos,
	   const uint8_t *key, size_t len_key,
	   const uint8_t *val0, size_t len_val0,
	   const uint8_t *val1, size_t len_val1,
	   uint8_t **merged_val, size_t *len_merged_val)
{
	user_func_merge(clos,
			key, len_key,
			val0, len_val0,
			val1, len_val1,
			merged_val, len_merged_val);
	count_merged += 1;
}

static void
merge(void)
{
	const uint8_t *key, *val;
	size_t len_key, len_val;
	struct mtbl_iter *it;

	it = mtbl_source_iter(mtbl_merger_source(merger));
	assert(it != NULL);
	while (mtbl_iter_next(it, &key, &len_key, &val, &len_val) == mtbl_res_success) {
		mtbl_res res = mtbl_writer_add(writer, key, len_key, val, len_val);
		assert(res == mtbl_res_success);
		if ((++count % STATS_INTERVAL) == 0)
			print_stats();
	}
	mtbl_iter_destroy(&it);
	mtbl_merger_destroy(&merger);
	mtbl_writer_destroy(&writer);
}

static void
init_dso(void)
{
	merge_dso_path = getenv("MTBL_MERGE_DSO");
	merge_dso_prefix = getenv("MTBL_MERGE_FUNC_PREFIX");

	if (merge_dso_path == NULL) {
		fprintf(stderr, "Error: MTBL_MERGE_DSO environment variable not set.\n\n");
		usage();
	}
	if (merge_dso_prefix == NULL) {
		fprintf(stderr, "Error: MTBL_MERGE_FUNC_PREFIX environment variable not set.\n\n");
		usage();
	}

	dlerror();
	void *handle = dlopen(merge_dso_path, RTLD_NOW);
	if (handle == NULL) {
		fprintf(stderr, "Error: dlopen() failed: %s\n", dlerror());
		exit(EXIT_FAILURE);
	}

	/*
	 * Note: the code below uses ubufs in a subtle way, three times.
	 * The first ubuf_append() with strlen() does not append a NUL.
	 * The second ubuf_append() with sizeof() does append a NUL.
	 * The ubuf_data() call accesses the underlying data as a
	 * NUL-terminated C string.
	 */

	/* merge func */
	ubuf *func_name = ubuf_init(0);
	ubuf_append(func_name,
		    (const uint8_t *) merge_dso_prefix,
		    strlen(merge_dso_prefix));
	ubuf_append(func_name,
		    (const uint8_t *) "_func",
		    sizeof("_func"));
	user_func_merge = dlsym(handle, (const char *) ubuf_data(func_name));
	if (user_func_merge == NULL) {
		fprintf(stderr, "Error: user merge function required but not found in DSO.\n\n");
		usage();
	}

	/* init func */
	ubuf_clip(func_name, 0);
	ubuf_append(func_name,
		    (const uint8_t *) merge_dso_prefix,
		    strlen(merge_dso_prefix));
	ubuf_append(func_name,
		    (const uint8_t *) "_init_func",
		    sizeof("_init_func"));
	user_func_init = dlsym(handle, (const char *) ubuf_data(func_name));
	if (user_func_init != NULL)
		user_clos = user_func_init();

	/* free func */
	ubuf_clip(func_name, 0);
	ubuf_append(func_name,
		    (const uint8_t *) merge_dso_prefix,
		    strlen(merge_dso_prefix));
	ubuf_append(func_name,
		    (const uint8_t *) "_free_func",
		    sizeof("_free_func"));
	user_func_free = dlsym(handle, (const char *) ubuf_data(func_name));

	/* cleanup */
	ubuf_destroy(&func_name);
}

static size_t
get_block_size(void)
{
	uint64_t sz;
	if (getenv_int("MTBL_MERGE_BLOCK_SIZE", &sz))
		return ((size_t) sz);
	return (DEFAULT_BLOCK_SIZE);
}

static void
init_mtbl(void)
{
	struct mtbl_merger_options *mopt;
	struct mtbl_writer_options *wopt;

	mopt = mtbl_merger_options_init();
	wopt = mtbl_writer_options_init();

	mtbl_merger_options_set_merge_func(mopt, merge_func, user_clos);
	mtbl_writer_options_set_compression(wopt, opt_compression_type);
	if (opt_compression_level != DEFAULT_COMPRESS_LEVEL)
		mtbl_writer_options_set_compression_level(wopt, opt_compression_level);
	mtbl_writer_options_set_block_size(wopt, opt_block_size);

	merger = mtbl_merger_init(mopt);
	assert(merger != NULL);

	fprintf(stderr, "%s: opening output file %s\n", program_name, mtbl_output_fname);
	writer = mtbl_writer_init(mtbl_output_fname, wopt);
	if (writer == NULL) {
		fprintf(stderr, "Error: mtbl_writer_init() failed.\n\n");
		usage();
	}

	mtbl_merger_options_destroy(&mopt);
	mtbl_writer_options_destroy(&wopt);
}

static bool
parse_long(const char *str, long int *val)
{
	char *endptr;

	errno = 0;
	*val = strtol(str, &endptr, 0);

	if ((errno == ERANGE && (*val == LONG_MAX || *val == LONG_MIN))
	    || (errno != 0 && *val == 0))
	{
		return false;
	}

	if (endptr == str || *endptr != '\0')
		return false;

	return true;
}

static bool
parse_arg_block_size(const char *arg)
{
	long int arg_block_size = 0;

	if (getenv("MTBL_MERGE_BLOCK_SIZE") != NULL) {
		fprintf(stderr,
			"%s: Error: specify block size via command-line or environment, not both\n",
			program_name);
		return false;
	}

	if (!parse_long(arg, &arg_block_size))
		return false;

	opt_block_size = arg_block_size;
	return true;
}

static bool
parse_arg_compression(const char *arg)
{
	mtbl_res res;

	res = mtbl_compression_type_from_str(arg, &opt_compression_type);
	return res == mtbl_res_success;
}

static bool
parse_arg_compression_level(const char *arg)
{
	char *endp;
	opt_compression_level = strtol(arg, &endp, 10);
	return *endp == '\0';
}

int
main(int argc, char **argv)
{
	setlocale(LC_ALL, "");
	program_name = argv[0];

	opt_block_size = get_block_size();

	int c;
	while ((c = getopt(argc, argv, "b:c:l:")) != -1) {
		switch (c) {
		case 'b':
			if (!parse_arg_block_size(optarg))
				usage();
			break;
		case 'c':
			if (!parse_arg_compression(optarg))
				usage();
			break;
		case 'l':
			if (!parse_arg_compression_level(optarg))
				usage();
			break;
		default:
			usage();
		}
	}

	if (argc - optind < 2)
		usage();
	mtbl_output_fname = argv[argc - 1];

	/* open user dso */
	init_dso();

	/* open merger, writer */
	init_mtbl();

	/* open readers */
	const size_t n_readers = argc - 1 - optind;
	struct mtbl_reader *readers[n_readers];
	for (size_t i = 0; i < n_readers; i++) {
		const char *fname = argv[i + optind];
		fprintf(stderr, "%s: opening input file %s\n", program_name, fname);
		readers[i] = mtbl_reader_init(fname, NULL);
		if (readers[i] == NULL) {
			fprintf(stderr, "Error: mtbl_reader_init() failed.\n\n");
			usage();
		}
		mtbl_merger_add_source(merger, mtbl_reader_source(readers[i]));
	}

	/* do merge */
	my_timespec_get(&start_time);
	merge();

	/* cleanup readers */
	for (size_t i = 0; i < n_readers; i++)
		mtbl_reader_destroy(&readers[i]);

	/* call user cleanup */
	if (user_func_free != NULL)
		user_func_free(user_clos);

	print_stats();

	return (EXIT_SUCCESS);
}
